zookeeper-metrics-providers/zookeeper-prometheus-metrics/src/main/java/org/apache/zookeeper/metrics/prometheus/PrometheusMetricsProvider.java

```diff
@@ -0,0 +1,295 @@
+        private final ConcurrentMap<String, PrometheusLabelledSummary> summarySets = new ConcurrentHashMap<>();
+
+        @Override
+        public MetricsContext getContext(String name) {
+            // no hierarchy yet
+            return this;
+        }
+
+        @Override
+        public Counter getCounter(String name) {
+            return counters.computeIfAbsent(name, (n) -> {
+                return new PrometheusCounter(n);
+            });
+        }
+
+        @Override
+        public void registerGauge(String name, Gauge gauge) {
+            if (gauge == null) {
+                // null means 'unregister'
+                gauges.remove(name);
+            } else {
+                gauges.put(name, gauge);
+            }
+        }
+
+        @Override
+        public Summary getSummary(String name, DetailLevel detailLevel) {
+            if (detailLevel == DetailLevel.BASIC) {
+                return basicSummaries.computeIfAbsent(name, (n) -> {
+                    if (summaries.containsKey(n)) {
+                        throw new IllegalArgumentException("Already registered a non basic summary as " + n);
+                    }
+                    return new PrometheusSummary(name, detailLevel);
+                });
+            } else {
+                return summaries.computeIfAbsent(name, (n) -> {
+                    if (basicSummaries.containsKey(n)) {
+                        throw new IllegalArgumentException("Already registered a basic summary as " + n);
+                    }
+                    return new PrometheusSummary(name, detailLevel);
+                });
+            }
+        }
+
+        @Override
+        public SummarySet getSummarySet(String name, DetailLevel detailLevel) {
+            if (detailLevel == DetailLevel.BASIC) {
+                return basicSummarySets.computeIfAbsent(name, (n) -> {
+                    if (summarySets.containsKey(n)) {
+                        throw new IllegalArgumentException("Already registered a non basic summary set as " + n);
+                    }
+                    return new PrometheusLabelledSummary(name, detailLevel);
+                });
+            } else {
+                return summarySets.computeIfAbsent(name, (n) -> {
+                    if (basicSummarySets.containsKey(n)) {
+                        throw new IllegalArgumentException("Already registered a basic summary set as " + n);
+                    }
+                    return new PrometheusLabelledSummary(name, detailLevel);
+                });
+            }
+        }
+
+    }
+
+    private class PrometheusCounter implements Counter {
+
+        private final io.prometheus.client.Counter inner;
+        private final String name;
+
+        public PrometheusCounter(String name) {
+            this.name = name;
+            this.inner = io.prometheus.client.Counter
+                    .build(name, name)
+                    .register(collectorRegistry)
```

Comment:
Why are you registering to both a custom and the registry?

Reply:
I will try to use directly the registry, but I feel I won't be able to implement all of the required features non-Prometheus related.
stay tuned